File: C++/reverseList.cpp
#include <iostream>
using namespace std;
// 只需要完成逆置链表函数
struct ListNode
{
    int val;
    struct ListNode *next;
    ListNode(int x) : val(x), next(NULL)
    {
    }
};

class Solution
{
public:
    ListNode *ReverseList(ListNode *pHead)
    {
        if (pHead == NULL || pHead->next == NULL)
            return pHead;
        ListNode *pReverseNode = ReverseList(pHead->next);
        pHead->next->next = pHead;
        pHead->next = NULL;

        return pReverseNode;
    }
};

File: LeetCode/LeetCode-Offer/reverseList.php
<?php
require '../vendor/autoload.php';

class ListNode
{
    var $val;
    var $next = NULL;
    function __construct($x)
    {
        $this->val = $x;
    }
}


function ReverseList($pHead)
{
    if (!$pHead || !$pHead->next) {
        return $pHead;
    }

    $head = ReverseList($pHead->next);
    $pHead->next->next = $pHead;
    $pHead->next = null;
    return $head;
}


function ReverseList2($pHead)
{
    if (!$pHead)
        return $pHead;
    $lastNode = null;
    $node = $pHead;
    $nextNode = $pHead->next;
    for (;;) {
        $node->next = $lastNode;
        if (!$nextNode)
            return $node;
        $lastNode = $node;
        $node = $nextNode;
        $nextNode = $nextNode->next;
    }
}

$head = new ListNode(1);
$head->next = new ListNode(2);
$head->next->next = new ListNode(3);
$head->next->next->next = new ListNode(4);
$head->next->next->next->next = new ListNode(5);

// dd($head);
dd(ReverseList(null));
